### PR TITLE
enable low-precision pipeline

### PR DESCRIPTION
**Pull request by jiqing-feng**
Hi @amyeroberts @Narsil .

As the previous PR #31444 mentioned, it could enable low-precision pipelines by converting the outputs to `float()`. I followed the codes in [here](https://github.com/huggingface/transformers/blame/main/src/transformers/pipelines/text_classification.py#L205-L210). Do you mind taking a review? Thx!

### CODE REVIEW

**Comment by aliencaocao** [comment]
Duplicate of https://github.com/huggingface/transformers/pull/31342 haha

**Comment by jiqing-feng** [comment]
> Related to #31342 but I dont quite get your changes - what exactly does it fix? When I tested all the pipelines in fp16 none of them had issues outputting logits

Hi @aliencaocao , FP16 works fine but bf16 is not acceptable for numpy, you can see:

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 1]
```diff
@@ -557,7 +557,11 @@ def postprocess(
         key = "logits" if self.type == "ctc_with_lm" else "tokens"
         stride = None
         for outputs in model_outputs:
-            items = outputs[key].numpy()
+            if self.framework == "pt" and outputs[key].dtype in (torch.bfloat16, torch.float16):
+                # To enable using fp16 and bf16
+                items = outputs[key].float().numpy()
```
Comment (SunMarc) [inline src/transformers/pipelines/automatic_speech_recognition.py:562]: I saw in the other PR that we are also doing this for float16 even when numpy supports it. Maybe we should only do the `upscasting` for `bfloat16` ? WDYT @amyeroberts

File: src/transformers/pipelines/image_classification.py [hunk 2]
```diff
@@ -182,8 +180,9 @@ def postprocess(self, model_outputs, function_to_apply=None, top_k=5):
             top_k = self.model.config.num_labels
 
         outputs = model_outputs["logits"][0]
-        if self.framework == "pt" and outputs.dtype in (torch.bfloat16, torch.float16):
-            outputs = outputs.to(torch.float32).numpy()
+        if self.framework == "pt":
+            # To enable using fp16 and bf16
+            outputs = outputs.float().numpy()
```
Comment (SunMarc) [inline src/transformers/pipelines/image_classification.py:185]: why did you remove the dtype check ?

File: src/transformers/pipelines/token_classification.py [hunk 3]
```diff
@@ -299,7 +299,12 @@ def postprocess(self, all_outputs, aggregation_strategy=AggregationStrategy.NONE
             ignore_labels = ["O"]
         all_entities = []
         for model_outputs in all_outputs:
-            logits = model_outputs["logits"][0].numpy()
+            if self.framework == "pt":
+                # To enable using fp16 and bf16
+                logits = model_outputs["logits"][0].float().numpy()
```
Comment (SunMarc) [inline src/transformers/pipelines/token_classification.py:304]: add dtype check

File: src/transformers/pipelines/zero_shot_classification.py [hunk 4]
```diff
@@ -239,7 +239,10 @@ def _forward(self, inputs):
     def postprocess(self, model_outputs, multi_label=False):
         candidate_labels = [outputs["candidate_label"] for outputs in model_outputs]
         sequences = [outputs["sequence"] for outputs in model_outputs]
-        logits = np.concatenate([output["logits"].numpy() for output in model_outputs])
+        if self.framework == "pt":
+            logits = np.concatenate([output["logits"].float().numpy() for output in model_outputs])
+        else:
+            logits = np.concatenate([output["logits"].numpy() for output in model_outputs])
```
Comment (SunMarc) [inline src/transformers/pipelines/zero_shot_classification.py:245]: Add dtype check and maybe split these steps for readability

**Review by SunMarc — COMMENTED** [review]
Thanks for addding this @jiqing-feng ! Left a few comments ! Nice tests =)

**Comment by jiqing-feng** [comment]
Hi @SunMarc. I have fixed all your comments; please review them. BTW, the failed tests are due to the connection error, not related to my changes.

File: src/transformers/pipelines/image_classification.py [hunk 5]
```diff
@@ -182,8 +182,9 @@ def postprocess(self, model_outputs, function_to_apply=None, top_k=5):
             top_k = self.model.config.num_labels
 
         outputs = model_outputs["logits"][0]
-        if self.framework == "pt" and outputs.dtype in (torch.bfloat16, torch.float16):
-            outputs = outputs.to(torch.float32).numpy()
+        if self.framework == "pt" and outputs.dtype == torch.bfloat16:
+            # To enable using fp16 and bf16
```
Comment (SunMarc) [inline src/transformers/pipelines/image_classification.py:186]:
```suggestion
            # To enable using bf16
```

File: src/transformers/pipelines/token_classification.py [hunk 6]
```diff
@@ -299,7 +301,12 @@ def postprocess(self, all_outputs, aggregation_strategy=AggregationStrategy.NONE
             ignore_labels = ["O"]
         all_entities = []
         for model_outputs in all_outputs:
-            logits = model_outputs["logits"][0].numpy()
+            if self.framework == "pt" and model_outputs["logits"][0].dtype == torch.bfloat16:
+                # To enable using fp16 and bf16
```
Comment (SunMarc) [inline src/transformers/pipelines/token_classification.py:305]:
```suggestion
                # To enable using bf16
```

File: src/transformers/testing_utils.py [hunk 7]
```diff
@@ -2083,7 +2083,7 @@ def nested_simplify(obj, decimals=3):
         return nested_simplify(obj.numpy().tolist())
     elif isinstance(obj, float):
         return round(obj, decimals)
-    elif isinstance(obj, (np.int32, np.float32)):
+    elif isinstance(obj, (np.int32, np.float32, np.float16)):
```
Comment (SunMarc) [inline src/transformers/testing_utils.py:2086]: Why do we need to change that ? The fp16 support was working in the past for pipeline. Is it because you are using it in your tests ?

**Review by SunMarc — COMMENTED** [review]
Thanks for iterating ! Just a few nits !

File: src/transformers/testing_utils.py (hunk 7 above)
Comment (jiqing-feng) [reply src/transformers/testing_utils.py:2086]: Yes, it is for the fp16 tests.

Comment (SunMarc) [reply src/transformers/testing_utils.py:2086]: sounds good !

**Review by SunMarc — APPROVED** [review]
Thanks for iterating @jiqing-feng  !

**Comment by jiqing-feng** [comment]
Hi @amyeroberts , would you please review this PR? Thx!

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . I have passed all tests, do you mind take a review and merge it? Thx!

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . This PR should be ready to merge, please take a review, thx!

**Comment by amyeroberts** [comment]
Hi @jiqing-feng, thanks for opening this PR! I'll get to reviewing it soon, but will likely be in a few days.

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . Do you mind reviewing this PR? Thx.

**Comment by jiqing-feng** [comment]
Hi @amyeroberts . Do you think it could be merged?

**Comment by yao-matrix** [comment]
@amyeroberts , could you help review this PR? Thx.

**Comment by yao-matrix** [comment]
@SunMarc , do you have a suggestion any other people from HF can help review and merge this PR? Seems @amyeroberts has no bandwidth on this these serval months. Thx.

**Comment by LysandreJik** [comment]
Sorry for the delay @yao-matrix!

@Rocketknight1 is managing the pipeline; Matt, would you mind reviewing this PR when you have a second?

File: src/transformers/pipelines/image_classification.py [hunk 8]
```diff
@@ -184,8 +184,9 @@ def postprocess(self, model_outputs, function_to_apply=None, top_k=5):
             top_k = self.model.config.num_labels
 
         outputs = model_outputs["logits"][0]
-        if self.framework == "pt" and outputs.dtype in (torch.bfloat16, torch.float16):
-            outputs = outputs.to(torch.float32).numpy()
+        if self.framework == "pt" and outputs.dtype == torch.bfloat16:
+            # To enable using bf16
+            outputs = outputs.float().numpy()
```
Comment (Rocketknight1) [inline src/transformers/pipelines/image_classification.py:189]: This seems like a regression to me! The previous code worked with both `torch.bfloat16` and `torch.float16`.

**Review by Rocketknight1 — APPROVED** [review]
Overall this PR seems good to me! However, I prefer `if outputs.dtype in (torch.bfloat16, torch.float16)` rather than `if outputs.dtype == torch.bfloat16`, so we can catch `float16` as well.

Other than that, I'm happy with it!

File: src/transformers/pipelines/image_classification.py (hunk 8 above)
Comment (jiqing-feng) [reply src/transformers/pipelines/image_classification.py:189]: I use torch 2.4 in python 3.10 and it failed:
image

**Comment by jiqing-feng** [comment]
> Overall this PR seems good to me! However, I prefer `if outputs.dtype in (torch.bfloat16, torch.float16)` rather than `if outputs.dtype == torch.bfloat16`, so we can catch `float16` as well.
> 
> Other than that, I'm happy with it!

Hi @Rocketknight1 , thanks for your review, I have fixed your comments.

**Comment by jiqing-feng** [comment]
Hi @Rocketknight1 , do you mind helping to re-run the tests?

**Comment by Rocketknight1** [comment]
@jiqing-feng tests pass now and this looks good - are you okay for us to merge it?

Also cc @LysandreJik for final review, but no rush!

**Comment by jiqing-feng** [comment]
> @jiqing-feng tests pass now and this looks good - are you okay for us to merge it?
> 

Yes, please.

**Review by LysandreJik — APPROVED** [review]
That works for me! Thank you @jiqing-feng!
